### PR TITLE
[autoscaler][local] Fix scale down

### LINKED ISSUE #17022: [autoscaler][local] Downscaling doesn't work on on-prem clusters
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

See title. Nodes are marked terminated in a state file, but ray stop is not executed.
Either modify LocalNodeProvider.terminate_nodes to run ray stop via command runner or explicitly disallow min_workers != max_workers.

Discussion:
https://github.com/ray-project/ray/issues/15459#issuecomment-878662828
https://github.com/ray-project/ray/issues/15459#issuecomment-878701730
https://github.com/ray-project/ray/pull/16202#discussion_r645119583

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Closes https://github.com/ray-project/ray/issues/17022

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
It's unfortunately inconvenient for me to test this directly, so someone else with convenient access to an on-prem cluster will have to try it out.

Testing strategy:
Use a local node provider config with min_workers = 0, max_workers = 2, idle_timeout_minutes=1, and the following setup commands:
```
setup_commands:
    - if [ $(which ray) ]; then pip uninstall ray -y; fi
    - pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"
    - if [ -d ray ]; then rm -Rf ray; fi
    - git clone --single-branch --branch fix-local-node-provider-scale-down https://github.com/DmitriGekhtman/ray
    - pushd ray/python/ray && python setup-dev.py -y && popd
```
Do anything on the Ray cluster that triggers scale up of the two workers. Verify that the workers are there by running ray status. Wait for scale-down and verify that workers are gone by running ray status.

File: python/ray/autoscaler/_private/commands.py [hunk 1]
```diff
@@ -470,7 +473,13 @@ def kill_node(config_file: str, yes: bool, hard: bool,
...
         cli_logger.print("No worker nodes detected.")
         return None
-    node = random.choice(nodes)
+    if not _node:
+        node = random.choice(nodes)
+    else:
+        # Kill specific node if supplied via hidden arg.
+        # Used exclusively by autoscaler to stop Ray on nodes of on-prem
+        # clusters.
+        node = _node
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/commands.py:482]: nit: can you flip this to be 
```
if _node:
  node = _node
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 2]
```diff
@@ -318,9 +313,7 @@ def _update(self):
                                        " Failed to update node."
                                        " Node has already been terminated.")
                 if nodes_to_terminate:
-                    self.prom_metrics.stopped_nodes.inc(
-                        len(nodes_to_terminate))
-                    self.provider.terminate_nodes(nodes_to_terminate)
+                    self._terminate_nodes_and_cleanup(nodes_to_terminate)
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/autoscaler.py:316]: Is it okay that we don't previously untrack nodes here?

**Review by ijrsvt — COMMENTED** [review]
Two quick comments!

File: python/ray/autoscaler/_private/autoscaler.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py:316]: Untrack has an if block that makes it okay to untrack twice.

Comment (ijrsvt) [reply python/ray/autoscaler/_private/autoscaler.py:316]: Great :)

File: python/ray/autoscaler/_private/commands.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/commands.py:482]: yeah good idea -- `if uncommon condition: do uncommon thing`

**Review by ijrsvt — APPROVED** [review]
LGTM now!

**Comment by DmitriGekhtman** [comment]
I need fix import bugs -- do not merge.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by DmitriGekhtman** [comment]
This is no longer relevant -- "draining" nodes now takes care of this.

**Pull request closed without merging**